tmt/utils.py

```diff
@@ -63,10 +71,36 @@
 SHELL_OPTIONS = 'set -eo pipefail'
 
 
+# A stand-in variable for generic use.
+T = TypeVar('T')
+
+ContextType = Dict[str, str]
+
+
+class BaseLoggerFnType(Protocol):
+    def __call__(
+            self,
+            key: str,
+            value: Optional[str] = None,
+            color: Optional[str] = None,
+            shift: int = 0,
+            level: int = 1,
+            err: bool = False) -> None:
+        pass
+
+
+class SemanticLoggerFnType(Protocol):
+    def __call__(self, message: str, shift: int = 0) -> None:
+        pass
+
+
+LoggerFnType = Union[BaseLoggerFnType, SemanticLoggerFnType]
```

Comment:
I may be missing it but I don't see any use of this `Union`. Is it some leftover, or what's the purpose of this?

Reply:
I *think* we shall be able to get rid of this, sooner or later. I encountered places where "a logging function" is called or referred to, and the problem is the signature of logging functions is not the same, there exist two types of logging functions. I wanted to capture this duality right away, but, 1. it may appear it's not going to be actually used because all logging functions ever assigned to variable/parameter are one of these types, never both, and 2. as we discussed recently, logging may see some changes Soon (TM).

So, yes, it's unused in `tmt/utils.py`, but I wanted to express the fact that when tmt code talks about "logging function", it means two signatures can be encountered. I'd prefer to drop this to just one type, e.g. by pronouncing `SemanticLoggerFnType` as the one true type of logging functions any code shall ever pass around or use, and `BaseLoggerFnType` would be reduced to one function somewhere inside the logging subsystem, called by various semantic logging functions with proper parameters. Then it'd be easy: drop `BaseLoggerFnType` because nobody is ever going to use it, nobody's going to pass this base implementation around, and rename `SemanticLoggerFnType` to `LoggerFnType`, replacing the union.

I expect to resolve this conundrum over time, by annotating places where logging functions are used as names with annotations, landing on some kind of solution I described above, because we discover logging functions of `BaseLoggerFnType` are never passed around. Right now, I can't guarantee that.